### DISCUSSION TITLE
Feedback-driven Roadmap

### DISCUSSION THREAD

**Discussion by dswbx** [bknd-io/bknd#2] [Ideas]
Hey everyone, happy to have you here!

I'm opening this topic to ask you what should be focused on next. To give you a head-start, I'm listing some thoughts, but you're not limited to them.

* Data
    * More fields (e.g. rich text)
    * Better relational data management
    * Vector fields/functions
    * Embedded replicas
    * Locale/offline writes
    * Advanced „with“ prop (à la drizzle)
    * UUID support as primary
    * OpenAPI editor
    * Entity features like sorting, hierarchy, timestamps, etc.
    * Data list views like kanban, calendar, etc.
    * Custom queries
    * Renaming fields and entities
    * Visual filter options
    * Immutable entities
* Auth
    * Cookie based auth
    * bcrypt algorithm
    * Session based tokens
    * Refresh tokens
    * More social OAuth providers
    * TOTP
    * MFA
    * External auth integration
    * Advanced permission control
    * Additional fields for sign ups
* Media
    * Better in-app viewers
    * Mime type and size checking on entity media fields
    * Chunked uploads
    * Folders
* New/misc features
    * Logs, analytics
    * Dashboard/reporting
    * Websockets/realtime
    * Notifications (email, sms, push)
    * Geo restriction
    * Rate limiting
* Misc
    * Improve docs
    * More tutorials, use cases

**Comment by cameronapak** [comment]
Hey, thanks for creating an Astro integration. I'm excited to test it for [Freedom Stack](https://freedom.faith.tools/)

**Reply by dswbx** [reply]
You are very welcome! It was actually by far the easiest framework to integrate. I just checked out your Freedom Stack, looks great! From what I can see consulting the astro db docs and your db setup script, bknd should integrate nicely since you have full db access. Just keep in mind that bknd expects entities/tables and fields to be created inside bknd, as it maintains its own metadata. But that's only a soft limit at the moment, left it open for the future to auto introspect.

Please don't hesitate to reach out if you run into any issues, I'm more than happy to help!

**Reply by cameronapak** [reply]
@dswbx I love Astro, and what you're creating fits perfectly well with the vision of Freedom Stack. Excited to dig in. Thank you!

**Comment by joakimbeng** [comment]
Do you consider local first/offline support? I haven't seen it mentioned anywhere and I think it's an important feature

**Reply by dswbx** [reply]
You‘re right, it isn‘t mentioned, but it should be! Will add it to the list, it‘s definitely an important feature. 

Mind to elaborate how you‘d expect it to behave (and in which environment)? This would help designing it based on realistic scenarios.

**Reply by joakimbeng** [reply]
My main use case would be a PWA app, so the browser as the environment.

Having the same API on the client side as the server would be the best. I imagine the client module to store data in a local database (maybe wasm-sqlite) first and then sync it with the backend layer.
There are a lot of alternatives in this area, like PowerSync, LiveStore, Electric, to mention a few, but I haven't found anyone that is an all inclusive solution that includes for instance auth. Having a single "backend as a service" solution that gives you everything you need is really compelling IMO! It seems that bknd is ticking off a lot of the boxes, and if you can get offline support as well it's a given!

**Reply by dswbx** [reply]
Thanks a lot for the details, this makes a lot of sense. How would you expect conflicts to be resolved?

I've screened the options you listed, and there is probably more (as you said). However, I really want to stay with libSQL so I don't have to maintain multiple databases of the same type (SQLite), so I'm afraid we have to wait until the Turso team has completed their offline reads/writes implementation. It's currently in private beta and focusing on server and mobile environments, hopefully browser will come soon as well 🤞 (see their [blog article](https://turso.tech/blog/introducing-offline-writes-for-turso))

In the meantime I'll explore alternatives, but investing into an external provider could be tough, as I hope Turso will catch up soon. 

But since bknd works in the browser, would fully offline also be an option?

**Reply by joakimbeng** [reply]
Great article you shared! It sounds very promising and I think their way for conflict resolution would work. I also found a repo of theirs experimenting with libSQL-wasm, which would make browser support for local-first possible, in a similar fashion that PowerSync does with SQLite-wasm.

I'll keep an eye on Turso's development!

With "fully offline" you mean no backend sync at all? That's not an option in my case as I want multi device support.
There's no rush though, I'm more in the investigation stage at the moment so nothing needs to exist right now.

**Reply by dswbx** [reply]
A core principle of bknd is to avoid dependency on a single database vendor (even though it's ironically depending on libSQL atm). Therefore offline capabilities should be provided by the vendor itself, or may be used in an isolated server-side (for remote) and client-side (for local) implementation. Currently only the "server-side" part is implemented using the `Connection` class which generally allows bknd to work with any [kysely supported dialect](https://kysely.dev/docs/dialects). Btw this approach extends to other areas like authentication, storage and server environment – aiming to prevent vendor limitations and lock-ins for the future at the cost of deeper integration with specific features. 

I've explored your suggestions and additional options from localfirstweb.dev, there are many great choices. However, the varying APIs make it challenging to generalize a solution at this stage. For bknd, the most promising direction might be to wait for Turso’s implementation, but I’ll continue my research. What has been your experience with any of these options?

Yes, I meant no remote backend at all. Expected that, but had to ask :)

**Reply by joakimbeng** [reply]
I've only recently started to delve deeper into this topic and am currently building an app using PowerSync together with Supabase.
Both the developer and user experience is great! I can use Kysely or plain SQL in the frontend and I get both offline support and live updates for the queries without any effort ☺️
The only thing is that you have to design your database a certain way to be able to write sync queries in PowerSync, because it uses a very limited set of the SQL language. But it seems that they are working on improving it with support for joins, which would make it much better and easier to write sync rules for an existing database.

Even though I really like the PowerSync + Supabase combo I would prefer if the syncing and offline/local-first support would be built-in into Supabase. Writing both sync rules (PowerSync) and RLS policies (Supabase) feels a bit like doing the same thing twice. If Supabase instead could sync a limited set of the database (e.g. what's requested by the client plus what the client has changed/added) and have a local database on the client, then you would only have to write the RLS policies. That or something in the same direction would be really nice 😃

**Comment by Vachan0** [comment]
Have you considered using [Better Auth](https://www.better-auth.com/docs/introduction) for authentication? It’s plugin-based, works with Hono and other frameworks, and might be a good fit for this project.

**Reply by dswbx** [reply]
I like better auth, but every dependency adds debt and potential limitations (along with benefits of course). I want the authentication part to be as deeply integrated as possible, and I can only guarantee this by not using a ready-made solution. Hope this makes sense.

**Comment by airtonix** [comment]
no plans for schema driven migrations? 

Seems like a massive oversight? 

Or if i wanted migrations, am I not the target market? 

How were you envisioning BKND working in different `dev`, `stg` and `prod` environments? 

The cli looks like it can export the schema, but I can't see anything about importing it into another env? 

And even more concerning is this: 

<img width="805" height="313" alt="image" src="https://github.com/user-attachments/assets/f46bf45d-ee02-46a4-9ad6-b4318f17094e" />

 I was dead set hyped about what you created here until it dawn upon me that i'd be condemning myself to a hilariously outrageous life of hell when i need to migrate changes from local to any other env...

**Reply by cameronapak** [reply]
I see. For example, if you make a bunch of changes in the GUI on local dev but want to have those reflected (migrated) into prod, you'd have to follow every single thing you did in the GUI on production. 

I see bknd as an SDK that has all batteries included: auth, db management, s3, even emails now with plugins. 

So, as a dev, migrations make sense. 

@dswbx created an issue https://github.com/bknd-io/bknd/issues/201 about migrations

**Reply by dswbx** [reply]
@airtonix thanks for reaching out. May I ask which assumptions made you hyped? Maybe it's not that far away, but in any case it would be interesting to hear about it.

bknd is different to other backends/frameworks in wish you control everything with code, that you can check into VCS and manage environments like that. It's UI first, and with that, there comes different concepts and obstacles. The main issue is that secrets (such as access credentials to a remote storage) are part of the configuration that you can enter in the UI. If everything was managed with code, you could simply extract them from environments variables.

As @cameronapak pointed out, there are already concepts floating around to enable environments/migrations. May I ask what you're explicitly looking for?

Controlling all behavior with code, meaning that you could not make adjustments in the UI?
Or would you want to make configurations visually using the UI in dev, and then push it to staging/production (which in turn cannot be changed in the UI)?
The main obstacles are secrets, those would need to be extracted before persisting, as you shouldn't check them into VCS. If there should be only a "code first" option which disables all UI configurations, that would be relatively easy – but misses the point of bknd imo.

Curious to hear about what people are looking for. There have been only a few requests in that direction, there are many ways to accomplish it, but I want to make sure it's implemented in a way people actually want.

**Reply by airtonix** [reply]
> @airtonix thanks for reaching out. May I ask which assumptions made you hyped? Maybe it's not that far away, but in any case it would be interesting to hear about it.

My backend expectations come from a history of using and enjoying Django. Except that I now loath to use python.

So to some degree you're almost there.
m

> bknd is different to other backends/frameworks in wish you control everything with code, that you can check into VCS and manage environments like that. It's UI first, and with that, there comes different concepts and obstacles. The main issue is that secrets (such as access credentials to a remote storage) are part of the configuration that you can enter in the UI. If everything was managed with code, you could simply extract them from environments variables.
> 
> As @cameronapak pointed out, there are already concepts floating around to enable environments/migrations. May I ask what you're explicitly looking for?
> 
> Controlling all behavior with code, meaning that you could not make adjustments in the UI? Or would you want to make configurations visually using the UI in dev, and then push it to staging/production (which in turn cannot be changed in the UI)?

Yes, it should be possible to distill any schema changes made in the UI to an export. 

With JSON Schema to AST, it should be very possible to generate a drizzle/kysley/whatever `.ts` file that i then include in my code base and bundle into an artifact (docker image) that gets promoted through my other environments.

> The main obstacles are secrets, those would need to be extracted before persisting, as you shouldn't check them into VCS.

100% secrets shouldn't be transferred. I personally think that while your solution of provide a secret service is nice, in this scenario I'm fairly sure people in my camp will probably want to not use your secret store and have secrets come from some where else (Vault, SSM, Docker Secrets, etc).

So right now I image your strategy for getting secrets in the db into the various things that need them is basically: 

```
getSecretValue(key):
  access bknd db > use secret
```

where it should be 

```
getSecretValue(key):
  access configured secret store > 
    bknd? > access bknd db >  use secret
    ssm? > fetch > use secret
    1password? > use opclient > use secret
    etc...
```

On the question of schema / data / migrations

If you approach this problem as schema migration vs data migrations, then we can simply have the schema data migration export unticked by default.

So: 

```
[ Export Schema and Data ]

Schema

[ x ] Users 
[ x ] Posts
[ x ] Secrets

Data 

[  ] Users 
[  ] Posts
[  ] Secrets 

```

> If there should be only a "code first" option which disables all UI configurations, that would be relatively easy – but misses the point of bknd imo.
> 
> Curious to hear about what people are looking for. There have been only a few requests in that direction, there are many ways to accomplish it, but I want to make sure it's implemented in a way people actually want.


My point is that your default experience should be a UI that uses this plugin system under the hood and is by default "simple" yet not easily transferrable to other environments.


Another way to ask this same question: 

- How am I going to recover from disasters? 
- How am I going to setup offline testing environments for my developers?

**Reply by airtonix** [reply]
It's a nice project, but I've decided through experience that there's going to be way too much drama in my life were I to base my entire app on bknd.

My main concerns are : 

- I want local development to be easy and fast for my entire stack. Relying on `bknd` in it's current state means i can't easily offline that and run it locally in a way that reflects one of my environments. (I'm not interested in running some kind of vm snapshot workflow here either)

- By design, there's only a PROD environment, which changes how people approach problem solving (for the worse). So i've already got cold shivers imaging the jank and spaghetti that will accumulate due to people trying to hold onto a gajillion backwards compatiable changes.

**Comment by donnimsipa** [comment]
Hi! I noticed “Custom queries” listed under the Data section as a potential upcoming feature.

I wanted to confirm whether this refers to the ability to run raw SQL queries (e.g., SELECT, INSERT, UPDATE, DELETE) directly through the BKND UI.

Is raw SQL execution planned as part of “Custom queries,” or is that intended to be a different feature altogether?

Thanks for clarifying!

**Reply by dswbx** [reply]
hey @donnimsipa, kind of both. It's already possible to use custom queries using the exposed kysely instance, like:

```bknd.config.ts
import type { BkndConfig } from "bknd";
import { sql } from "kysely";

export default {
  onBuilt: async (app) => {
    const db = app.connection.kysely;
    app.server.get("/custom_query", async (c) => {
      return c.json(await db.selectFrom("pages").selectAll().execute());
    });
    app.server.get("/custom_query2", async (c) => {
      return c.json(await sql`select * from pages`.execute(db));
    });
  }
} satisfies BkndConfig;
```

"Custom queries" in this roadmap refers to adding custom "views" to entities based on the query object using the Admin UI. E.g. you could be fetching like this:

```ts
await api.data.readMany("posts", {
  where: {
    views: { $gt: 100 }
  },
  sort: "-views"
});
```

So it would be nice to have a way to save this query (say "with_100_views"), and then have it added as an endpoint, e.g. `/api/data/entity/posts/query/with_100_views` which returns the same result as the manual example above. This would also allow to display custom queries as tabs to the Admin UI.

Hope this clarifies it!
